core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2823,73 +2914,313 @@ public void cleanUp() {
+            // save for later
+            errors.add(e);
         }
-        // TODO: how to wait for the completion of the last job?
-        Trigger.timer = null;
+    }
 
-        Timer.shutdown();
+    private void _cleanUpShutdownTimer(List<Throwable> errors) {
+        LOGGER.log(Level.FINE, "Shutting down timer");
+        try {
+            Timer.shutdown();
+        } catch (SecurityException e) {
+            LOGGER.log(WARNING, "Not permitted to shut down Timer", e);
+            errors.add(e);
+        } catch (OutOfMemoryError e) {
+            // we should just propagate this, no point trying to log
+            throw e;
+        } catch (LinkageError e) {
+            LOGGER.log(SEVERE, "Failed to shut down Timer", e);
+            // safe to ignore and continue for this one
+        } catch (Throwable e) {
+            LOGGER.log(SEVERE, "Failed to shut down Timer", e);
+            // save for later
+            errors.add(e);
+        }
+    }
 
-        if(tcpSlaveAgentListener!=null)
-            tcpSlaveAgentListener.shutdown();
+    private void _cleanUpShutdownTcpSlaveAgent(List<Throwable> errors) {
+        if(tcpSlaveAgentListener!=null) {
+            LOGGER.log(FINE, "Shutting down TCP/IP slave agent listener");
+            try {
+                tcpSlaveAgentListener.shutdown();
+            } catch (OutOfMemoryError e) {
+                // we should just propagate this, no point trying to log
+                throw e;
+            } catch (LinkageError e) {
+                LOGGER.log(SEVERE, "Failed to shut down TCP/IP slave agent listener", e);
+                // safe to ignore and continue for this one
+            } catch (Throwable e) {
+                LOGGER.log(SEVERE, "Failed to shut down TCP/IP slave agent listener", e);
+                // save for later
+                errors.add(e);
+            }
+        }
+    }
 
-        if(pluginManager!=null) // be defensive. there could be some ugly timing related issues
-            pluginManager.stop();
+    private void _cleanUpShutdownPluginManager(List<Throwable> errors) {
+        if(pluginManager!=null) {// be defensive. there could be some ugly timing related issues
+            LOGGER.log(Level.INFO, "Stopping plugin manager");
+            try {
+                pluginManager.stop();
+            } catch (OutOfMemoryError e) {
+                // we should just propagate this, no point trying to log
+                throw e;
+            } catch (LinkageError e) {
+                LOGGER.log(SEVERE, "Failed to stop plugin manager", e);
+                // safe to ignore and continue for this one
+            } catch (Throwable e) {
+                LOGGER.log(SEVERE, "Failed to stop plugin manager", e);
+                // save for later
+                errors.add(e);
+            }
+        }
+    }
 
-        if(getRootDir().exists())
+    private void _cleanUpPersistQueue(List<Throwable> errors) {
```

Comment:
Well if you do a `System.exit(0)` or `kill -9` then there is not much we can do... but yes, normal shutdown should keep the queue